### DISCUSSION TITLE
Operation consistency model

### DISCUSSION THREAD

**Discussion by uroni** [minio/minio#14339] [Q&A] [answered]
When I write to a minio bucket, what are the guarantees I have once the operation is finished?

I did sift a bit through the source code and found no calls to either fsync or fssync.

This is also something that should be added prominently to the documentation. E.g. "Minio is a distributed object storage server with eventual consistency". I may also have missed that bit.

**Comment by harshavardhana** [comment, marked as answer]
MinIO changed the way it does write for many different situations over the years 

- https://github.com/minio/minio/blob/master/cmd/xl-storage.go#L1762
- https://github.com/minio/minio/blob/master/cmd/xl-storage.go#L1794
- https://github.com/minio/minio/blob/master/cmd/xl-storage.go#L1565

### Write phase
- All writes for `xl.meta` that carries metadata about erasure coding, user metadata, etc write in O_SYNC for all their data 
- All writes for the data part are written in aligned writes with O_DIRECT and then perform a Fdatasync() on the file descriptor at the end.

### Read phase
O_DIRECT is also enabled by default for all READs as well.

- When reads are not aligned O_DIRECT is not used however page cache is freed. https://github.com/minio/minio/blob/master/cmd/xl-storage.go#L1715
- Optionally you can turn off O_DIRECT for reads
